Fix inconsistency between Python and JIT power operation

szewaiyuen7 opened the pull request:
When base is zero and exponent is negative, python power operation would raise a run time error while JIT would return inf. 
This is due to underlying cmath pow(a, b) function in some compilers would return inf instead of throwing error. Therefore, we are wrapping cmath pow function inside a wrapper and perform an addition check for zero base with negative exponent in the wrapper.

Note: Check that lint failure is not related to files changed in this PR.

Fixes #62556

szewaiyuen7 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -912,7 +921,7 @@ RegisterOperators reg(
            int64_t a, b;
            pop(stack, a, b);
-           push(stack, pow(a, b));
+           push(stack, powWrapper(a, b));
```
For the case of `int a ** int b` in Python, `DEFINE_GENERIC_OP_WITH_COMPLEX` is used to resolve the aten::pow. I am not sure how to test for this operator `aten::pow.int_to_int(int a, int b) -> int` as I am not sure how to trigger this in Python

zhxchen17 replied:
Yeah, that seems to be the case. Since our overall goal is to align the two different behaviors, I'd suggest to just replace all pow call to the wrapper call, and then I'll figure out whether they're used later and do something in a separate diff.

szewaiyuen7 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -893,13 +902,13 @@ RegisterOperators reg(
-     DEFINE_INT_FLOAT_OP(aten::pow, pow(a, b), float),
+     DEFINE_INT_FLOAT_OP(aten::pow, powWrapper(a, b), float),
      DEFINE_FLOAT_COMPLEX_OP(aten::pow, pow(a, b), complex),
      DEFINE_SCALAR_BINARY_OP_AVOID_COLLISION(
```
I am not sure how to trigger the `DEFINE_SCALAR_BINARY_OP_AVOID_COLLISION` from Python. What Python type correspond to a scalar?

zhxchen17 replied:
same, I don't know how to test this right now either, but since `Scalar` is just union of all cases you've tested, it won't be an issue to just replace all `pow` calls.

zhxchen17 commented on test/jit/test_aten_pow.py:
```diff
@@ -0,0 +1,87 @@
+        def fn_int_int(a: int, b: int):
+            return a ** b
+        # Existing correct behavior of aten::pow
+        assert fn_int_int(2, 1) == 2 ** 1
```
self.assertEq() is preferred over `assert`

gmagogsfm commented:
@zhxchen17 @szewaiyuen7 Note that this change is actually BC-breaking, please watch out for internal breakages after importing to phabricator.

szewaiyuen7 committed:
Fix inconsisteny between Python and JIT power operation

ghstack-source-id: 60b68cd4b9a98af3e9bd04e3f93edfe541661931
Pull Request resolved: https://github.com/pytorch/pytorch/pull/62842

zhxchen17 commented:
seems like we have close this in favor of https://github.com/pytorch/pytorch/pull/62842 .
